### PR TITLE
DialogService samples

### LINKED ISSUE #51: [Update] 07 DialogService
### Sample

07-DialogService

#### Reason for Update

This sample is currently stubbed and needs actual implementation.

TODO

- Add a basic dialog that shows passing of dialog parameters via QueryString and DialogParameters
- Add basic Dialog that has a form that controls the return value of CanCloseDialog

### PR DESCRIPTION
**Pull request by hnabbasi**
# Dialog Service

fixes #51 

The new Prism Dialog Service is designed to provide you a minimalistic framework that you can build on to provide richly styled dialogs that match the look and feel of your app.

|Samples|Selector Dialog|Tap to close
--|--|--
 |  | 

## Dialog View
Let's start with the view for our dialog.

```xml
<ContentView xmlns="http://xamarin.com/schemas/2014/forms"
    xmlns:x="http://schemas.microsoft.com/winfx/2009/xaml"
    xmlns:prism="http://prismlibrary.com"
    prism:DialogLayout.RelativeWidthRequest="0.75" 
    ...>
    
    <!-- The View -->

</ContentView>
```
## Dialog ViewModel
Dialog view's ViewModel implements `IDialogAware`
```csharp
public class TermsDialogViewModel : BindableBase, IDialogAware 
{ 
    public TermsDialogViewModel()
    {
        Title = "Terms and Conditions";
        SubmitCommand = new DelegateCommand(OnSubmitTapped)
            .ObservesCanExecute(() => CanContinue);
        CancelCommand = new DelegateCommand(OnCancelTapped);
    }

    public DelegateCommand SubmitCommand { get; }
    public DelegateCommand CancelCommand { get; }

    // IDialogAware
    public event Action<IDialogParameters> RequestClose;
    public bool CanCloseDialog() { ... };
    public void OnDialogClosed() { }
    public void OnDialogOpened(IDialogParameters parameters) { }
}
```
## Dialog Registration
Register your dialogs to the container in RegisterTypes() of the App class
```csharp
containerRegistry.RegisterDialog<TermsDialog>();
containerRegistry.RegisterDialog<LockingDialog>();
containerRegistry.RegisterDialog<NameDialog>();
containerRegistry.RegisterDialog<ContactSelectorDialog>();
```
## Usage
```csharp
private readonly IDialogService _dialogService;
public SelectorPageViewModel(IDialogService dialogService)
{
    _dialogService = dialogService;
}
```
Show the dialog,
```csharp
_dialogService.ShowDialog("LockingDialog");
```
Show the dialog and pass in `DialogParameters`,
```csharp
// using Query string
_dialogService.ShowDialog("LockingDialog?Question=Can navigate away?");

// using DialogParameters
_dialogService.ShowDialog("LockingDialog", new DialogParameters
{
    { "Question", "Can navigate away?" }
});
```
Show the dialog and handle callback,
```csharp
_dialogService.ShowDialog("TermsDialog", new DialogParameters(), HandleCallback());

private void HandleCallback(IDialogResult result)
{...}
```
## Extensions
DialogService can be extended to meet the needs. In this sample, the DialogService has a couple of `async` extension methods and a generic extension.
extensions
### ShowDialogAsync
This extensions allows showing of the dialog to be awaited.
```csharp
public static Task<IDialogResult> ShowDialogAsync(this IDialogService dialogService,
    string name,
    IDialogParameters parameters)
{
    var tcs = new TaskCompletionSource<IDialogResult>();

    try
    {
        dialogService.ShowDialog(name, parameters, (result) => {
            if (result.Exception != null)
            {
                tcs.SetException(result.Exception);
                return;
            }
            tcs.SetResult(result);
        });
    }
    catch (Exception ex)
    {
        tcs.SetException(ex);
    }
    return tcs.Task;
}
```
usage,
```csharp
var r = await _dialogService.ShowDialogAsync("NameDialog");
return r.Parameters.GetValue<string>("Name");
```

### ShowDialogAsync`<T>`
This extension shows the dialog and expects a parameter with the name matching the `Type` name.
```csharp
public static Task<T> ShowDialogAsync<T>(this IDialogService dialogService,
            string name)
{
    var tcs = new TaskCompletionSource<T>();

    try
    {
        dialogService.ShowDialog(name, (result) =>
        {
            if (result.Exception != null)
            {
                tcs.SetException(result.Exception);
                return;
            }
            tcs.SetResult(result.Parameters.GetValue<T>(typeof(T).Name));
        });
    }
    catch (Exception ex)
    {
        tcs.SetException(ex);
    }
    return tcs.Task;
}
```
usage,
```csharp
var contact = await _dialogService.ShowDialogAsync<Contact>("ContactSelectorDialog");
```
>Note: These extensions are for reference purposes only. Use these techniques to make extensions that fit your needs.

### CODE REVIEW

File: 07-DialogService/src/PrismSample/PrismSample.csproj [hunk 1]
```diff
@@ -7,6 +7,10 @@
 
   <ItemGroup>
     <PackageReference Include="Prism.DryIoc.Forms" Version="7.2.0.1422" />
-    <PackageReference Include="Xamarin.Forms" Version="4.3.0.991211" />
+    <PackageReference Include="Xamarin.Forms" Version="4.4.0.991265" />
+  </ItemGroup>
+  <ItemGroup>
+    <Folder Include="Dialogs\" />
+    <Folder Include="Helpers\" />
   </ItemGroup>
```
Comment (dansiegel) [inline 07-DialogService/src/PrismSample/PrismSample.csproj]: not actually needed even though VS will automatically put it in when you create the folder... you can get rid of these...

File: 07-DialogService/src/PrismSample/PrismSample.csproj [hunk 2]
```diff
@@ -7,6 +7,10 @@
 
   <ItemGroup>
     <PackageReference Include="Prism.DryIoc.Forms" Version="7.2.0.1422" />
-    <PackageReference Include="Xamarin.Forms" Version="4.3.0.991211" />
+    <PackageReference Include="Xamarin.Forms" Version="4.4.0.991265" />
```
Comment (dansiegel) [inline 07-DialogService/src/PrismSample/PrismSample.csproj]: no updates please for any existing references.

File: 07-DialogService/src/PrismSample/ViewModels/SelectorPageViewModel.cs [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+            get { return _contact; }
+            set { SetProperty(ref _contact, value); }
+        }
+
+        public DelegateCommand GetContactCommand { get; }
+        
+        async void OnGetContactTapped()
+        {
+            Contact = await _dialogService.ShowDialogAsync<Contact>("ContactSelectorDialog");
+        }
```
Comment (dansiegel) [inline 07-DialogService/src/PrismSample/ViewModels/SelectorPageViewModel.cs]: be sure to specify the accessor

File: 07-DialogService/src/PrismSample/Views/SelectorPage.xaml.cs [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+using Xamarin.Forms;
+using Xamarin.Forms.Xaml;
+
+namespace PrismSample.Views
+{
+    [XamlCompilation(XamlCompilationOptions.Compile)]
```
Comment (dansiegel) [inline 07-DialogService/src/PrismSample/Views/SelectorPage.xaml.cs]: already enabled at the assembly level

File: 07-DialogService/src/PrismSample.Android/PrismSample.Android.csproj [hunk 5]
```diff
@@ -56,7 +55,7 @@
     <PackageReference Include="Prism.DryIoc.Forms">
       <Version>7.2.0.1422</Version>
     </PackageReference>
-    <PackageReference Include="Xamarin.Forms" Version="4.3.0.991211" />
+    <PackageReference Include="Xamarin.Forms" Version="4.4.0.991265" />
```
[the remark this answers is not in the document]
Comment (hnabbasi) [reply 07-DialogService/src/PrismSample.Android/PrismSample.Android.csproj]: Done

File: 07-DialogService/src/PrismSample.Android/Resources/values/styles.xml [hunk 6]
```diff
@@ -11,12 +11,12 @@
...
     <!-- Set theme colors from https://aka.ms/material-colors -->
     <!-- colorPrimary is used for the default action bar background -->
-    <item name="colorPrimary">#2196F3</item>
+    <item name="colorPrimary">#00A5F2</item>
     <!-- colorPrimaryDark is used for the status bar -->
     <item name="colorPrimaryDark">#1976D2</item>
     <!-- colorAccent is used as the default value for colorControlActivated
          which is used to tint widgets -->
-    <item name="colorAccent">#FF4081</item>
+    <item name="colorAccent">#00A5F2</item>
```
[the remark this answers is not in the document]
Comment (hnabbasi) [reply 07-DialogService/src/PrismSample.Android/Resources/values/styles.xml]: Done

File: 07-DialogService/src/PrismSample.iOS/PrismSample.iOS.csproj [hunk 7]
```diff
@@ -123,12 +123,12 @@
     <PackageReference Include="Prism.DryIoc.Forms">
       <Version>7.2.0.1422</Version>
     </PackageReference>
-    <PackageReference Include="Xamarin.Forms" Version="4.3.0.991211" />
+    <PackageReference Include="Xamarin.Forms" Version="4.4.0.991265" />
```
[the remark this answers is not in the document]
Comment (hnabbasi) [reply 07-DialogService/src/PrismSample.iOS/PrismSample.iOS.csproj]: Done

File: 07-DialogService/src/PrismSample/Dialogs/ContactSelectorDialog.xaml.cs [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+{
+    [XamlCompilation(XamlCompilationOptions.Compile)]
+    public partial class ContactSelectorDialog : ContentView
+    {
+        public ContactSelectorDialog()
+        {
+            InitializeComponent();
+        }
+    }
+}
```
[the remark this answers is not in the document]
Comment (hnabbasi) [reply 07-DialogService/src/PrismSample/Dialogs/ContactSelectorDialog.xaml.cs]: Done

File: 07-DialogService/src/PrismSample/Dialogs/TermsDialog.xaml [hunk 9]
```diff
@@ -0,0 +1,53 @@
+﻿<?xml version="1.0" encoding="UTF-8"?>
+<ContentView xmlns="http://xamarin.com/schemas/2014/forms"
+             xmlns:x="http://schemas.microsoft.com/winfx/2009/xaml"
+             xmlns:prism="http://prismlibrary.com"
+             prism:DialogLayout.RelativeWidthRequest="0.75"
+             x:Class="PrismSample.Views.TermsDialog">
+  <Frame CornerRadius="20">
```
[the remark this answers is not in the document]
Comment (hnabbasi) [reply 07-DialogService/src/PrismSample/Dialogs/TermsDialog.xaml]: Done

File: 07-DialogService/src/PrismSample/Dialogs/NameDialog.xaml [hunk 10]
```diff
@@ -0,0 +1,15 @@
+﻿<?xml version="1.0" encoding="UTF-8"?>
+<ContentView xmlns="http://xamarin.com/schemas/2014/forms"
+             xmlns:x="http://schemas.microsoft.com/winfx/2009/xaml"
+             xmlns:prism="http://prismlibrary.com"
+             prism:DialogLayout.RelativeWidthRequest="0.75"
+             x:Class="PrismSample.Views.NameDialog">
+    <ContentView.Content>
+      <Frame CornerRadius="20">
```
[the remark this answers is not in the document]
Comment (hnabbasi) [reply 07-DialogService/src/PrismSample/Dialogs/NameDialog.xaml]: Done

File: 07-DialogService/src/PrismSample/Dialogs/LockingDialog.xaml [hunk 11]
```diff
@@ -0,0 +1,20 @@
+﻿<?xml version="1.0" encoding="UTF-8"?>
+<ContentView xmlns="http://xamarin.com/schemas/2014/forms"
+             xmlns:x="http://schemas.microsoft.com/winfx/2009/xaml"
+             x:Class="PrismSample.Views.LockingDialog"
+             xmlns:prism="http://prismlibrary.com"
+             prism:DialogLayout.CloseOnBackgroundTapped="{Binding CloseOnTap}" >
+    <ContentView.Content>
+      <Frame CornerRadius="20">
```
[the remark this answers is not in the document]
Comment (hnabbasi) [reply 07-DialogService/src/PrismSample/Dialogs/LockingDialog.xaml]: Done

File: 07-DialogService/src/PrismSample/Dialogs/TermsDialog.xaml.cs [hunk 12]
```diff
@@ -0,0 +1,14 @@
...
+
+namespace PrismSample.Views
+{
+    public partial class TermsDialog : ContentView
+    {
+        public TermsDialog()
+        {
+            InitializeComponent();
+        }
+    }
```
[the remark this answers is not in the document]
Comment (hnabbasi) [reply 07-DialogService/src/PrismSample/Dialogs/TermsDialog.xaml.cs]: Done

File: 07-DialogService/src/PrismSample/Dialogs/TermsDialogViewModel.cs [hunk 13]
```diff
@@ -0,0 +1,70 @@
...
+            CancelCommand = new DelegateCommand(OnCancelTapped);
+        }
+
+        public DelegateCommand SubmitCommand { get; }
+        public DelegateCommand CancelCommand { get; }
+
+        void OnSubmitTapped()
+        {
+            RequestClose(new DialogParameters { { "accepted", true } });
+        }
```
[the remark this answers is not in the document]
Comment (hnabbasi) [reply 07-DialogService/src/PrismSample/Dialogs/TermsDialogViewModel.cs]: Done

File: 07-DialogService/src/PrismSample/PrismSample.csproj (hunk 1 above)
Comment (hnabbasi) [reply 07-DialogService/src/PrismSample/PrismSample.csproj]: Done

File: 07-DialogService/src/PrismSample/PrismSample.csproj (hunk 2 above)
Comment (hnabbasi) [reply 07-DialogService/src/PrismSample/PrismSample.csproj]: Done

File: 07-DialogService/src/PrismSample/ViewModels/SelectorPageViewModel.cs (hunk 3 above)
Comment (hnabbasi) [reply 07-DialogService/src/PrismSample/ViewModels/SelectorPageViewModel.cs]: Done

File: 07-DialogService/src/PrismSample/App.xaml.cs [hunk 14]
```diff
@@ -1,22 +1,20 @@
...
-        public App()
-        {
-        }
-
         protected override async void OnInitialized()
         {
             InitializeComponent();
 
-            var result = await NavigationService.NavigateAsync("MainPage");
+            var result = await NavigationService.NavigateAsync($"{nameof(NavigationPage)}/{nameof(MainPage)}");
```
[the remark this answers is not in the document]
Comment (hnabbasi) [reply 07-DialogService/src/PrismSample/App.xaml.cs]: Done

File: 07-DialogService/src/PrismSample/Dialogs/ContactSelectorDialogViewModel.cs [hunk 15]
```diff
@@ -0,0 +1,46 @@
...
+        private Contact _selectedContact;
+        public Contact SelectedContact
+        {
+            get { return _selectedContact; }
+            set
+            {
+                SetProperty(ref _selectedContact, value);
+                if(value != null)
+                    RequestClose(new DialogParameters {{"Contact", value}});
+            }
```
[the remark this answers is not in the document]
Comment (hnabbasi) [reply 07-DialogService/src/PrismSample/Dialogs/ContactSelectorDialogViewModel.cs]: Nice. done.

File: 07-DialogService/src/PrismSample/Views/SelectorPage.xaml.cs (hunk 4 above)
Comment (hnabbasi) [reply 07-DialogService/src/PrismSample/Views/SelectorPage.xaml.cs]: Done

File: 07-DialogService/src/PrismSample/Dialogs/ContactSelectorDialogViewModel.cs [hunk 16]
```diff
@@ -0,0 +1,46 @@
...
+        public ContactSelectorDialogViewModel(IContactsService contactsService)
+        {
+            _contactsService = contactsService;
+        }
+        
+        private List<Contact> _contacts;
+        public List<Contact> Contacts
+        {
+            get { return _contacts; }
+            set { SetProperty(ref _contacts, value); }
```
[the remark this answers is not in the document]
Comment (hnabbasi) [reply 07-DialogService/src/PrismSample/Dialogs/ContactSelectorDialogViewModel.cs]: Updated my snippets. Done.

File: 07-DialogService/src/PrismSample.Android/Resources/values/styles.xml [hunk 17]
```diff
@@ -11,7 +11,7 @@
     <item name="windowActionBar">false</item>
     <!-- Set theme colors from https://aka.ms/material-colors -->
     <!-- colorPrimary is used for the default action bar background -->
-    <item name="colorPrimary">#2196F3</item>
+    <item name="colorPrimary">#00A5F2</item>
```
Comment (dansiegel) [inline 07-DialogService/src/PrismSample.Android/Resources/values/styles.xml]: no style changes please

Comment (hnabbasi) [reply 07-DialogService/src/PrismSample.Android/Resources/values/styles.xml]: Damn it. It should be reverted properly now 👍

**Comment by dansiegel** [comment]
looks great, thanks for the PR @hnabbasi

**Pull request merged**